### PR TITLE
WIP: Scene

### PR DESCRIPTION
**Pull request by cgphilipp**

### CODE REVIEW

File: examples/scene-example.ts [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+        return true;
+    }
+
+    /**
+     * Uninitializes Buffers, Textures, and Program.
+     */
+    protected onUninitialize(): void {
+        super.uninitialize();
+
+        // TODO: make sure that all meshes and programs inside of the scene get cleaned
```
Comment (scheibel) [inline examples/scene-example.ts:129]: Which components needs to be uninitialized?
Are all member variables that are initialized during `onInitialize` required to be undefined?

File: examples/scene-example.ts [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+     * changed or the renderer was invalidated @see{@link invalidate}.
+     * Updates the navigaten and the AntiAliasingKernel.
+     * @returns whether to redraw
+     */
+    protected onUpdate(): boolean {
+        const gl = this.context.gl;
+        // Resize
+        if (this._altered.frameSize) {
+            this._camera.viewport = [this._frameSize[0], this._frameSize[1]];
+            gl.viewport(0, 0, this._frameSize[0], this._frameSize[1]);
```
Comment (scheibel) [inline examples/scene-example.ts]: I'm not sure if a render pass should rely on the global OpenGL viewport. I think it is more stable to assure the correct size before each draw call.

File: examples/scene-example.ts [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        }
+        if (this._altered.canvasSize) {
+            this._camera.aspect = this._canvasSize[0] / this._canvasSize[1];
+        }
+
+        // Update clear color
+        if (this._altered.clearColor) {
+            const c = this._clearColor;
+            gl.clearColor(c[0], c[1], c[2], c[3]);
```
Comment (scheibel) [inline examples/scene-example.ts]: Same here as for OpenGL viewport.

File: examples/scene-example.ts [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    }
+
+    /**
+     * @todo comment
+     * @param frameNumber - for intermediate frames in accumulation rendering.
+     */
+    protected onFrame(frameNumber: number): void {
+        const gl = this._context.gl;
+
+        this._camera.viewport = [this._frameSize[0], this._frameSize[1]];
```
Comment (scheibel) [inline examples/scene-example.ts]: This is also updated during `onUpdate`. Why doing it twice?

File: examples/scene-example.ts [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+
+        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
```
Comment (scheibel) [inline examples/scene-example.ts]: May a renderer assume it doesn't have to share OpenGL with other renderers? For this specific example: the clear method assumes the default framebuffer is still bound.

File: examples/scene-example.ts [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        this._forwardPass.frame();
+
+        gl.cullFace(gl.BACK);
+        gl.disable(gl.CULL_FACE);
+    }
+
+    /**
+     * @todo comment ...
+     */
+    protected onSwap(): void {
```
Comment (scheibel) [inline examples/scene-example.ts:191]: Just for comment: The expected functionality is to ensure the contents of a frame to be on the OpenGL backbuffer. Is this example renders into the backbuffer, nothing has to be done. The swap of front buffer and back buffer is scheduled after the invocation of this function.

File: examples/scene-example.ts [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+        const aMeshVertex = program.attribute('a_vertex', 0);
+        const aMeshTexCoord = program.attribute('a_texcoord', 1);
+        geometry.initialize(aMeshVertex, aMeshTexCoord);
+
+        const sphereComponent = new GeometryComponent();
+        sphereComponent.geometry = geometry;
+        sphereComponent.material = material;
+
+        node.addComponent(sphereComponent);
```
Comment (scheibel) [inline examples/scene-example.ts]: What is the target domain of `Component`? For the domains of a scene graph, a renderer, or even geometry, I think this term is too unspecific.

File: package.json [hunk 8]
```diff
@@ -66,7 +66,7 @@
     },
     "scripts": {
         "build": "webpack-cli --mode development && node pugconfig.js",
-        "build-watch": "webpack-cli --mode production --watch",
+        "build-watch": "webpack-cli --mode development --watch",
```
Comment (scheibel) [inline package.json:69]: Is this an intentional change? Was it wrong before?

File: source/core/icosahedron.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+    /**
+     * Generates texture coordinates for the icosahedron.
+     */
+    generateTextureCoordinates(): void {
+        /* Create array for the texture coordinates. */
+        this._texcoords = new Array<vec2>();
+
+        /* Generate texture coordinates. */
+        const ofs = 0.5;
```
Comment (scheibel) [inline source/core/icosahedron.ts:182]: Please qualify `ofs` by either extending the variable name of extending the comment.

File: source/core/resourcemanager.ts [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        if (resource instanceof Texture2D) {
+            const texture = resource as Texture2D;
+            if (this._texture2Ds.has(texture.identifier)) {
+                return false;
+            }
+            this._texture2Ds.set(texture.identifier, texture);
+            return true;
+        }
+
+        // if (resource instanceof ...) {
```
Comment (scheibel) [inline source/core/resourcemanager.ts:70]: Does this have to remain?

File: source/scene/forwardscenerenderpass.ts [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+    frame(): void {
+        assert(this._target && this._target.valid, `valid target expected`);
+        if (this._scene === undefined) {
+            return;
+        }
+
+        const gl = this._context.gl;
+
+        const size = this._target.size;
+        gl.viewport(0, 0, size[0], size[1]);
```
Comment (scheibel) [inline source/scene/forwardscenerenderpass.ts:95]: The viewport is updated here...I don't think the renderer above should assume its control over the OpenGL viewport.

File: source/scene/material.ts [hunk 12]
```diff
@@ -0,0 +1,21 @@
+
+import { Bindable } from '../bindable';
+import { Program } from '../program';
+
+export abstract class Material implements Bindable {
```
Comment (scheibel) [inline source/scene/material.ts:5]: `Material` is not an OpenGL concept but I think `Bindable` was introduced for OpenGL concepts, mainly. What is the take on this?

File: source/scene/scenenode.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+
+    /**
+     * @todo comment
+     * @param node - @todo comment
+     */
+    addNode(node: SceneNode): SceneNode {
+        node._parent = this;
```
Comment (scheibel) [inline source/scene/scenenode.ts:56]: Is it possible to add a scene node twice?

File: source/scene/scenenode.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+        return component;
+    }
+
+    /**
+     * @todo comment
+     * @param type - @todo comment
+     */
+    componentsOfType(type: string): Array<SceneNodeComponent> {
+        return this._components.filter((component: SceneNodeComponent) => type === component.type);
```
Comment (scheibel) [inline source/scene/scenenode.ts:77]: No subtyping planned?

File: examples/scene-example.ts (hunk 6 above)
Comment (cgcostume) [reply examples/scene-example.ts:191]: added parts of the suggested comments to the render interface ... will be merged at some point, not part of this pr

File: package.json (hunk 8 above)
Comment (cgcostume) [reply package.json:69]: i approve

File: source/core/box.ts [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+        if (this._textured) {
+            this._buffers[1].attribDisable(1, true, true);
+        }
+    }
+
+    /**
+     * Creates the vertex buffer object (VBO) and creates and initializes the buffer's data store.
+     * @param aVertex - Attribute binding point for vertices.
+     */
+    initialize(aVertex: GLuint, aTexCoord: GLuint): boolean {
```
Comment (cgcostume) [inline source/core/box.ts]: please refer to ``NDCfillingtriangle``  - see initialize default attributes + getter vor attribute location queries.

File: source/core/icosahedron.ts [hunk 16]
```diff
@@ -0,0 +1,212 @@
+
+import { vec2, vec3 } from 'gl-matrix';
+import { v3 } from '../gl-matrix-extensions';
+
+type ivec3 = [number, number, number];
+
+/**
+ * Helper class to generate a sphere geometry based on a refinable icosahedron
+ */
+export class Icosahedron {
```
Comment (cgcostume) [inline source/core/icosahedron.ts:10]: extend ```Geometry```

File: source/core/icosahedron.ts [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+
+        /* Return index of the new vertex. */
+        return index;
+    }
+
+    /**
+     * Generates the geometry of the icosahedron.
+     * @param levels - Number of levels of refinement.
+     */
+    generateGeometry(levels: number): void {
```
Comment (cgcostume) [inline source/core/icosahedron.ts:163]: this could be initialize function (when extending geometry)

File: source/core/icosahedron.ts [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+        /* Refine geometry. */
+        if (levels > 0) {
+            this.refine(this._vertices, this._faces, Math.min(Math.max(levels, 0), 8));
+        }
+    }
+
+    /**
+     * Generates texture coordinates for the icosahedron.
+     */
+    generateTextureCoordinates(): void {
```
Comment (cgcostume) [inline source/core/icosahedron.ts:177]: per flag in initialize `... (..., generateTexCoords: boolean = true);` - and make function protected or include within `initialize`

File: source/core/icosahedron.ts (hunk 9 above)
Comment (cgcostume) [reply source/core/icosahedron.ts:182]: 😄

File: source/core/sphere.ts [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+            this._buffers[2].attribDisable(1, true, true);
+        }
+    }
+
+
+    /**
+     * Creates the vertex buffer object (VBO) and creates and initializes the buffer's data store.
+     * @param aVertex - Attribute binding point for vertices.
+     */
+    initialize(aVertex: GLuint, aTexCoord: GLuint): boolean {
```
Comment (cgcostume) [inline source/core/sphere.ts]: define defaults .. again see `ndcfillingtriangle`for reference design

File: source/scene/material.ts (hunk 12 above)
Comment (cgcostume) [reply source/scene/material.ts:5]: passt

**Review by cgcostume — APPROVED** [review]

File: source/core/icosahedron.ts (hunk 16 above)
Comment (cgphilipp) [reply source/core/icosahedron.ts:10]: This is actually just a helper class to initialize `Sphere` and does not represent a drawable geometry itself

**Pull request closed without merging**